community/cypher/cypher-compiler-2.3/src/main/scala/org/neo4j/cypher/internal/compiler/v2_3/planner/logical/steps/triadicSelectionFinder.scala

```diff
@@ -27,43 +27,83 @@ import org.neo4j.cypher.internal.frontend.v2_3.ast._
 import org.neo4j.cypher.internal.frontend.v2_3.ast.Expression
 
 object triadicSelectionFinder extends CandidateGenerator[LogicalPlan] {
-  override def apply(in: LogicalPlan, qg: QueryGraph)(implicit context: LogicalPlanningContext): Seq[LogicalPlan] = in match {
+
+  override def apply(in: LogicalPlan, qg: QueryGraph)(implicit context: LogicalPlanningContext): Seq[LogicalPlan] =
+    qg.selections.patternPredicatesGiven(in.availableSymbols).collect {
```

Comment:
What we are doing is restructuring an already solved plan into new possible solutions, including the solved predicates. So, for example, the pattern (a)-[:X]->(b)-[:X]->(c) has the solved predicate requiring the two relationships to be different. We retain that in the new plan as well. Predicates related to finding (b) are tacked onto the LHS of the triadic selection, ones related to finding (c) are currently tacked onto the output of triadic selection, but could probably be on the RHS of triadic selection (eg. labels on (c) or relationship uniqueness). I think the reason we put them on the outside of triadic is because they might include other predicates not related to triadic, which we don't want to fiddle with.